### ISSUE TITLE
Couple struct/ustruct to LPN (for cardiac mechanics)

### ISSUE THREAD

**Issue by aabrown100-git** [SimVascular/svFSIplus#73]
Recreating the issue found here https://github.com/SimVascular/svFSI/issues/72. 

The implementation of this feature for svFSI is complete and waiting for approval before merging. I will start translating this feature into C++ code for svFSIplus and perform the work here: https://github.com/aabrown100-git/svFSIplus/tree/0D_coupling.

**Comment by ktbolt** [comment]
@aabrown100-git what's the status of this?

**Comment by aabrown100-git** [comment]
I will get started on this next week.

**Comment by aabrown100-git** [comment]
In the first stage of this, I will make the required changes for 3D structure - 0D coupling without "capping". Capping is complicated, and the current implementation can be improved. To make the changes, I will compare https://github.com/vvedula22/svFSI with https://github.com/SimVascular/svFSI. There are additional features in https://github.com/vvedula22/svFSI not related to 3D-0D coupling, which will be translated in a future pull request. The files that contain changes relevant to 3D-0D coupling are

- [ ] ALLFUN.f
- [ ] BAFINI.f
- [ ] EQASSEM.f
- [ ] FLUID.f (comments only)
- [ ] LHSA.f (comments only)
- [ ] LS.f (comments only)
- [ ] MAIN.f (comments only)
- [ ] MOD.f (comments only)
- [ ] NN.f
- [ ] PIC.f (comments only)
- [ ] SETBC.f
- [ ] STRUCT.f (comments only)
- [ ] USTRUCT.f (comments only)
- [ ] VTKXML.f (comments only)
- [ ] ADDBCMUL.f (comments only)
- [ ] BC.f
- [ ] NSSOLVER.f (comments only)

**Comment by ktbolt** [comment]
@aabrown100-git Could you please give a high-level overview describing what changes you needed to make and how these relate to what was described in https://github.com/SimVascular/svFSI/issues/72.

This makes it easier to associate the components of a design with implementation.

**Comment by aabrown100-git** [comment]
Yes, in this first stage (coupling without capping), I need to
- Modify the calculation of flowrate (computed as the velocity flux integral over a surface) to use the deformed configuration of the structure.
- Modify the 0D contribution to the tangent matrix to (again) use the deformed configuration of the structure.

In the second stage (with capping), both of these calculations must be modified to account for a cap surface.

**Comment by aabrown100-git** [comment]
First stage changes are translated. Now trying to come up with a good test.

**Comment by aabrown100-git** [comment]
Created a simple test: an LV model coupled to a constant flowrate LPN. LV model is fixed at base, so that it doesn't need to be capped. Results are nearly identical between https://github.com/aabrown100-git/svFSIplus/tree/3D_0D_coupling and https://github.com/vvedula22/svFSI.

Now, cleaning up and adding some checks.

**Comment by aabrown100-git** [comment]
Just submitted pull request for these stage 1 changes. Next, will work on stage 2 changes, namely adding support for a "cap" surface to close a 0D-coupled surface. This requires modifications to the flowrate and tangent contribution calculations, as well as some code to process a user-provided cap mesh.

**Comment by ktbolt** [comment]
@aabrown100-git Has this work been completed ? If so then close this Issue. If not then please update the Issue, it's been a year since the last update.

**Comment by dseyler** [comment]
@ktbolt I am currently working on translating this feature to C++. I intend to compare the results to https://github.com/vvedula22/svFSI for test cases similar to those in the struct/ustruct folders with robin rather than dirichlet BCs on the base.

**Comment by dseyler** [comment]
Committed initial changes affecting 22 files. Will begin testing using the following test case: [tests/cases/struct/LV_NeoHookean_passive_sv0D_capped](https://github.com/dseyler/svMultiPhysics/tree/3D0DCouplingCap/tests/cases/struct/LV_NeoHookean_passive_sv0D_capped)

**Comment by dseyler** [comment]
The branch successfully compiles and processes a modified .xml file that includes a virtual capping .vtp mesh. Currently debugging to resolve crashes.

Note: The CopyBC and AddCapBC functions, implemented in readfiles.cpp, would be more appropriately structured as methods of the bcType class. For now, they remain as free functions, but future restructuring may improve readability.

**Comment by ktbolt** [comment]
Holy cow, you changed 22 files ? This is why we need classes and all that to localize changes.

**Comment by dseyler** [comment]
Test case passes in serial but crashes in parallel. Trying to compare outputs to [https://github.com/vvedula22/svFSI](url) but it fails in parallel as well due to issues with mesh partitioning with ParMETIS. Has anyone encountered similar issues with older branches of svFSI?

**Comment by ktbolt** [comment]
@dseyler I could not run [tests/cases/struct/LV_NeoHookean_passive_sv0D_capped](https://github.com/dseyler/svMultiPhysics/tree/3D0DCouplingCap/tests/cases/struct/LV_NeoHookean_passive_sv0D_capped)
- svFSIplus.xml file has the wrong format
- the mesh/mesh-surfaces/endo_cap.vtp does not have GlobalElementID data

What platform are you running on ?

**Comment by dseyler** [comment]
@ktbolt I've actually been using a different test case recently. Let me clean up my code and I'll update that in a bit!

As for the platform, I've been running on Apple M3 macOS 15.4.1

**Comment by ktbolt** [comment]
If the the **end_cap** mesh doesn't have a **GlobalElementID** then it is not a face and should not be included as an `Add_face` parameter. 

The **end_cap** mesh should maybe be included in the `Add_BC` parameter like so

```
   <Add_BC name="endo" >
      <Type> Neu </Type>
      <Time_dependence> Coupled </Time_dependence>
      <Follower_pressure_load> true </Follower_pressure_load>
      <Capping_face> mesh/mesh-surfaces/endo_cap.vtp  </Capping_face>    
   </Add_BC> 
```

**Comment by dseyler** [comment]
Fixed several bugs and the branch now successfully runs the test case [tests/cases/struct/LV_NeoHookean_passive_genBC_capped](https://github.com/dseyler/svMultiPhysics/tree/3D0DCouplingCap/tests/cases/struct/LV_NeoHookean_passive_genBC_capped) in serial and parallel with results in agreement with [SimVascular/svFSI/issues/72](https://github.com/SimVascular/svFSI/issues/72).

Next steps:

- Verify active contraction example on HPC cluster with docker container [tests/cases/struct/104-final-sim-9-25](https://github.com/dseyler/svMultiPhysics/tree/3D0DCouplingCap/tests/cases/struct/104-final-sim-9-25).
- Find a method to add virtual surfaces to LPNS using svZeroDSolver.

@ktbolt I believe this branch should be nearly ready for a pull request. Given that this is such an extensive feature, maybe it would be a good idea to find some time to review the changes.

**Comment by ktbolt** [comment]
@dseyler A good idea to review the changes before the pull request.

First provide here an overview of what you did with links to the appropriate code.

You and I can then interactively go through your changes together with any Team members that are interested.

**Comment by dseyler** [comment]
Below is an overview of the changes to each file in [my fork](https://github.com/dseyler/svMultiPhysics/tree/3D0DCouplingCap). The coupling scheme is nearly identical to that implemented by @aabrown100-git  in [SimVascular/svFSI/issues/72](https://github.com/SimVascular/svFSI/issues/72).

 

Code/Source/liner_solver/add_bc_mul.cpp
- Included capping surface contribution to S (v dot X).

Code/Source/liner_solver/bc.cpp
- Set virtual flag for lhs

Code/Source/liner_solver/fils_struct.hpp
- added member variables for virtual flag and index of capping face

Code/Source/liner_solver/fsils_api.hpp
- set default value of virtual flag to false.

Code/Source/solver/ComMod.h
- Added member variables for index of cap BC associated with a BC, cap name, and virtual flag.

Code/Source/solver/Parameters.cpp
- set parameters for “Capping_face” and “Virtual”

Code/Source/solver/all_fun.cpp
- implemented several changes so integrals over virtual faces are performed only on the main processor.

Code/Source/solver/baf_ini.cpp 
- Reproduced match_lhs_face which finds the lhs.face index corresponding to iM and iFa. 
- Used this function to share information with lhs.face if any face in msh.fa is capped.
- Skip virtual faces in bc_ini

Code/Source/solver/cmm.cpp
- Minor changes related to adding const cmMod argument to gnnb function.

Code/Source/solver/distribute.cpp
- Virtual faces do not lie on the volume mesh, so global element IDs and partitioning needed to be handled separately, as described in part_faceV

Code/Source/solver/eq_assem.cpp
- Compute contribution of resistance BC to tangent matrix using deformed geometry
- Skip elements on virtual faces (Ac = -1)

Code/Source/solver/load_msh.cpp
- Read virtual flag
- Assign global element IDs for virtual faces. Since virtual faces do not lie on the volume mesh, we set gE = 0

Code/Source/solver/nn.cpp
- Surface normals for virtual faces are computed in gnnbv

Code/Source/solver/read_files.cpp
- Read cap name
- If an LPN-coupled face has a cap, addCapBC copies boundary condition info (copyBC) from the capped BC and creates one for the cap face. It may make sense to restructure these as functions inside the bcType class in the future.

Code/Source/solver/set_bc.cpp
- Adds velocity flux from cap if face is capped.
- Sets resistance for capping BC to be the same as capped BC resistance

Code/Source/solver/vtk_xml.cpp
- Ignore “no element IDs” warning for virtual faces.

Code/Source/solver/vtk_xml_parser.cpp
- Set global element IDs to 0 for virtual faces

**Comment by ktbolt** [comment]
@dseyler Thank you for your detailed discussion ! Really just an overview of the idea behind the changes is required, like I needed to add contributions to boundary conditions from the cap or something.

There are lots of changes here so It would be easier for me to look at the code if you provided links to your changes. Or we could do it interactively.

**Comment by aabrown100-git** [comment]
@javijv4 have you pushed changes to https://github.com/javijv4/svMultiPhysics/tree/dev_CAP? I don't see your latest development.

**Comment by aabrown100-git** [comment]
@javijv4 Ah okay I think we're on the same page. I think there are 3 use cases for this 0D boundary condition class: fluid Neumann, fluid Dirichlet, and solid Neumann (which requires capping). And yes I agree I think ZeroDBoundaryCondition should handle all of these.

**Comment by javijv4** [comment]
Hey all, sorry it took so long, but I finally got the capped simulations working in parallel. I checked the results using the available tests and comparing the old svZeroD implementation with the new one and GenBC. For the capped simulations, I used @dseyler's GenBC implementation. I ran all tests using 1, 2, and 4 cores. 

### Test 1: LV_NeoHookean_passive_sv0D (No Cap)
<img width="500" alt="Image" src="https://github.com/user-attachments/assets/e9220c86-b45d-4d58-8cb0-94a8fe9e49e2" />

### Test 2: LV_NeoHookean_passive_sv0D_cap (GenBC from @dseyler)
<img width="500" alt="Image" src="https://github.com/user-attachments/assets/c5af9e2f-8e4b-49ef-ad3a-466ee822e4c3" />

### Test 3: pipe_RCR _sv0D
<img width="500" alt="Image" src="https://github.com/user-attachments/assets/383be81f-d65b-4e6d-8e07-823197c08d2d" />


Some general implementation notes:
1. The new class `ZeroDBoundaryCondition.cpp` handles all the things related to a coupled 3D-0D boundary condition, such as the integration of the coupled surfaces. 
2. `svZeroD_subroutines.cpp` handles the communication between the 3D and 0D by calling the ZeroDBoundaryCondition
3. Cap surfaces are associated with this BC, and their integration is handled by the class.

To call the new BC, this should be added
```
   <Add_BC name="coupled_surface" > 
      <Type> Neu </Type> 
      <Time_dependence> ZeroD_Coupled </Time_dependence> 
      <Follower_pressure_load> true </Follower_pressure_load>
      <svZeroDSolver_block> LV_IN </svZeroDSolver_block>
      <svZeroDSolver_cap> mesh/mesh-surfaces/endo_cap.vtp </svZeroDSolver_cap>  <!-- Optional to load the cap -->
   </Add_BC> 
```

Notes:
1. I only implemented coupled Neumann BCs. The implementation of Dirichlet coupled BCs is not working on the previous svZeroD version "coupled" BC, and, as far as I understand, its implementation is a bit more complicated, but I believe it will be easier to implement within the class. 
2. I left the old implementation intact for now. If `<Time_dependence> Coupled <Time_dependence>` is used, then the code will use the old implementation. Note that using a cap with the old svZeroD implementation is not possible.

My first goal was to make it work, so now I need to review the class implementation, clean it up, and see if some of the functions can be improved. I had to reimplement a bunch of face-related functions (such as reading the VTK inputs) in the new class because the ones used for Face handling needed inputs that were not available within the class. I will try now to see if I can make it work by using functions available elsewhere in the code, so there is less code duplication. 

@mrp089, the code is updated in my branch in case you want to test it out.

**Comment by mrp089** [comment]
Very cool! Thank you so much for your work on this, @javijv4! @federicaninno, can you give this a try?

**Comment by federicaninno** [comment]
I will, @mrp089! Thank you so much @javijv4!

**Comment by javijv4** [comment]
Hey all, I went over the implementation, cleaned it up, and optimized it as best I could. 

The class contains methods to:
1. Load a cap face .vtp: If a cap file path is given, this is called within the constructor. It reuses the functions from `VtkData.cpp` when possible (`load_cap_face_vtp`).
2. Integrate over a face (and associated cap if needed). The integration is called from `svZeroD_subroutines.cpp`. When a cap is present, the cap node positions are updated using their global node IDs, then the integration is performed and added to the associated face integral (`integrate_over_cap`).
3. Calculate the contribution to the surface normal integral from the cap (`compute_cap_valM`). This info is copied to the associated face pointer (`copy_cap_data_to_linear_solver_face`) such that we have access to it in the linear solver (see `add_vc_mul.cpp`).
4. Handle the parallelization: All the integrations are performed in the master rank. At the beginning, it distributes information such as whether each rank knows whether it is a Neumann or Dirichlet BC, or if there's an associated cap (`distribute`). If there is a cap, before the integration, the displacements for all cap nodes are gathered (`prepare_cap_gathered_data`) such that the master can perform the integration.

Here's a list of the files modified and why it is needed.
1. `add_bc_mul.cpp`: adds cap face contribution to S
2. `fils_struc.hpp`: included arrays for cap calculations
3. `precond.cpp`: adds precondition to the cap matrix
4. `ComMod.h`: adds ZeroDBoundaryCondition object
5. `Parameters.cpp`: added cap option
6. `baf_ini.cpp`: initializes the face integrations
7. `consts.h`: adding related bc_zerod items
8. `distribute.cpp` calls `zerod_bc `distribute function
9. `eq_assem.cpp`: copy data from zerod_bc to corresponding face
10. `read_files.cpp`: calls ZeroDBoundaryCondition constructor
11. `set_bc.cpp`: calls the flow rate calculation.
The branch also includes a modified `VtkData.cpp` related to #520, but I am hoping to include those changes there.

This implementation is in my [branch](https://github.com/javijv4/svMultiPhysics/tree/dev_CAP). 

@ktbolt, @aabrown100-git, @dseyler, could you take a look at the implementations (especially `ZeroDBoundaryCondition.cpp`) and let me know if it is acceptable?

@federicaninno did you get to test it? Is it working as expected?

@ncdorn, if you can run a test of one of your problems using my branch, that'd be great. Also, regarding #518, I can leave a template indicating where this should be implemented within the class. I am trying to replicate the current sv0D coupling (which accepts only Neumann BCs) in an object-oriented implementation and to add functionality to include a cap. If there's a code and a working example of a Dirichlet-coupled BC, I can work on including it; otherwise, I will leave that for the other issue.

**Comment by ktbolt** [comment]
@javijv4 An excellent overview of the implementation ! 

As a casual observer I have lost the context of this Issue over the 2+ years of comments I see here. I think that the values computed on the cap face are used to set the values used for a boundary condition for a given face. Is that correct ?

**Comment by javijv4** [comment]
@ktbolt, thanks! And yes, exactly. This is in the context of a coupled 3D-0D problem, where the pressure on a face of the 3D mesh (in case of a Neumann bc) is given by the solution of the 0D system. 

We want to calculate the flow rates $Q$ using the 3D velocities as,

$$Q = \int_{\Gamma_h} \mathbf u \cdot \mathbf n d\Gamma_h$$

where $u$ is the velocity and $n$ is the surface normal of the closed surface $\Gamma_h$. 
The flow rate $Q$ is passed to the 0D problem, which solves for the pressure $p$, which is applied as a Neumann BC $\mathbf h=p\cdot \mathbf n$. 

This is straightforward in fluids because the closed surface $\Gamma_h$ is the boundary of the 3D mesh. In solids, however, we care about the flow rate defined by the domain inside the solid geometry. In the case of a ventricle, for example, the domain is defined by the endocardial surface and the "cap" that closes it, which is not part of the 3D mesh.

$$Q = \int_{\Gamma_{endo}} \mathbf u \cdot \mathbf n d\Gamma_{endo} + \int_{\Gamma_{cap}} \mathbf u \cdot \mathbf n d\Gamma_{cap}$$

This "cap" integration is what was not implemented before. In the new implementation, all operations required for a coupled 3D-0D Neumann bc are handled by the new class, which also implements the routines needed to include this "cap".

**Comment by ktbolt** [comment]
@javijv4 Thank you for the clarification !

**Comment by federicaninno** [comment]
Hi @javijv4, I haven't tested it yet unfortunately but I am planning to do it next week.

Thank you so much

**Comment by ktbolt** [comment]
@javijv4 I spent a good amount of time going through your code; quite a bit of work there and mostly things look good !

 I do have some comments that I will add here so others can see them and then you and I can sit down and do an in-person code review if that sounds good to you.

This development is only for cardiac mechanics it seems so we should name things (e.g., classes, parameters) with that in mind to make it clear what they are used for. 

The `ZeroDBoundaryCondition` class name is confusing; this is what `svZeroD_subroutines.cpp` should maybe be called. Using the `svZeroDSolver` is just one way this can be used to set values for its associated face; it is also perhaps possible to do this using experimental data or some other computation (e.g., 1D solver). This should be renamed something with `cardiac` and `interface` in it I think.

As for the  XML parameters
```
   <Add_BC name="endo" >
      <Type> Neu </Type>
      <Time_dependence> ZeroD_Coupled </Time_dependence>
      <Follower_pressure_load> true </Follower_pressure_load>
      <svZeroDSolver_block> LV_IN </svZeroDSolver_block>
      <svZeroDSolver_cap> mesh/mesh-surfaces/endo_cap.vtp </svZeroDSolver_cap>
   </Add_BC>
```
`ZeroD_Coupled` is confusing. Just use `Coupled` and have a`Cardiac_interface` sub-section used to set the parameters used only for that
```
   <Add_BC name="endo" >
      <Type> Neu </Type>
      <Time_dependence> Coupled </Time_dependence>
      <Follower_pressure_load> true </Follower_pressure_load>
      
      <Cardiac_interface>
          <svZeroDSolver_block> LV_IN </svZeroDSolver_block>
         <Chamber_cap_surface> mesh/mesh-surfaces/endo_cap.vtp </Chamber_cap_surface>
       </Cardiac_interface>

   </Add_BC>
```
In future we may want to add more parameters so it is better to group parameters for context. 

It was not clear what `cap integration` was until I looked at the code. Note that you can use any surface for `Chamber_cap_surface`; I used epi.vtp and top.vtp and the solver ran to completion; perhaps that is by design ?

**Comment by javijv4** [comment]
@ktbolt, yes, in-person code review sounds good!

About the naming convention, you are right. I think this class should handle any BC where the values are set by another solver or external data, and the ZeroD name is not general enough. However, even though the cap is specific for cardiac applications (it might have other applications, but I can't think of any), the rest is not cardiac-specific. For example, this can be used in vascular problems to couple the fluid solution to the svZeroDSolver. So maybe we can call it CoupledBoundaryCondition or something like that? I do like the use of `interface` for `svZeroD_subroutines.cpp`.  So, just to propose a naming convention:
* `ZeroDBoundaryCondition.cpp` --> `CoupledBoundaryCondition.cpp`
* `svZeroD_subroutines.cpp` --> `svZeroD_interface.cpp`
* `svOneD_subroutines.cpp` --> `svOneD_interface.cpp` (when this exists)
* `Cardiac_interface` --> `Coupled_interface`
Curious to know what people think about this!

I will rename `ZeroD_Coupled` to `Coupled` and add a subsection for coupling-specific parameters. 

About using any surface for `Chamber_cap_surface`, that is a good point. I can add a check that ensures `cap_face` shares nodes with the BC `face`, which will prevent the user from using a cap that is not connected to the BC face.

**Comment by ktbolt** [comment]
@javijv4 The file renaming looks good !

**Comment by aabrown100-git** [comment]
@javijv4 Can we remove the old implementation of svZeroD coupling now? Also, can we remove cplBC support and migrate the genBC implementation into your CoupledBoundaryCondition class? I'm just thinking it would be good to consolidate all 0D/1D coupling code into your new class, as much as possible.

**Comment by alisonmarsden** [comment]
Does this also address the need for Dirichlet coupling raised in issue #518 ?   I'm still wondering whether @menon-karthik may have already addressed this with the coronary simulations he's running?   Dirichlet coupling (e.g. at a model inlet face) is working in genBC but not with svZeroD solver, as far as I know.

**Comment by javijv4** [comment]
@alisonmarsden, right now, this only refactors what it was already working with the coupling with svZeroD, and adds the cap support. So no, no Dirichlet. However, I think the new class should also handle the Dirichlet conditions. I am thinking of merging this with only Neumann, then working on adding support for Dirichlet BCs to address #518 and #517.

**Comment by alisonmarsden** [comment]
@javijv4 ok this sounds like a good plan to me.    We should do all of this in a coordinate manner - I agree.

**Comment by javijv4** [comment]
I finished modifying `CoupledBoundaryCondition.cpp` based on @ktbolt suggestions. @ktbolt, if you can take a look, that'd be great.

I am now working on updating the rest of the code to remove the old svZeroD coupling implementation and ensure the remaining BC-related code is as understandable as possible. One challenge related to #524 is that there is a single `cplBC` object in `ComMod`. I was initially thinking of removing the `cplBC` entirely and relying solely on the class, but this object contains information related to other coupling types, such as RCR, GenBC, etc. Plus, `cplBC` has the `svzerod_solver_interface`. So I am inclined to leave the `cplBC` object as untouched as possible so as not to break other things.

I also plan to test it in cases with two faces (like the example from @taeoukkim) to make sure it can handle more complex scenarios.

**Comment by ktbolt** [comment]
@javijv4 Your code is looking good ! 

It would be good to create a `CappingSurface` exception class for all of the throws you are doing; would simplify the code a bit.

**Comment by javijv4** [comment]
Hi all, I have finished cleaning up the rest of the code. I deleted all the old implementation and renamed everything to follow the naming convention as previously discussed. 

All tests are passing, including the tests with two outlets. I did two tests with this geometry, one with the same resistance on both sides (left) and one with different values on each side (right). Both give the same results as the legacy coupling implementation. 

<img width="400"  alt="Bifurcation with same outlet resistance" src="https://github.com/user-attachments/assets/b389b590-af0f-4e3e-9fde-2b4ecc5ceab7" />
<img width="400"  alt="Bifurcation with different outlet resistance" src="https://github.com/user-attachments/assets/309c981a-ae5b-4549-98d8-042e00988941" />

@ktbolt I added the CappingSurface exception class.

Next steps are:
* Refactor GenBC to also use CoupledBoundaryCondition
* Figure out how to make RCR work in combination with svZeroD (#524)
* Add the Dirichlet implementation (#518)

@aabrown100-git, @ktbolt do you think I can PR what I already implemented and tackle the other problems as part of the other issues? I think I have a pretty good idea of how to deal with the first two, but I have already changed a lot of things, and I want to make sure the current changes are acceptable as they are now before modifying the code more.

**Comment by taeoukkim** [comment]
Hi @javijv4 

As we discussed previously, I have been developing the 3D–1D coupling code based on your commit "I had broken some tests; bug fixed" in April 8th., and a substantial portion of the implementation is now complete.

At the mean time, I was able to resolve the issue regarding “Figure out how to make RCR work in combination with svZeroD (#524 )”. In addition, I have implemented the Dirichlet coupling for the 3D–1D framework (although I have not yet tested it with the 0D solver).

At this point, I think it would be best to some how merge your cleaned-up version of the code with my current implementation and then proceed with the remaining tasks.

I’m happy to discuss this further—perhaps we can set up a quick Zoom meeting to go over the details.

**Comment by aabrown100-git** [comment]
@javijv4 This looks great, yes I think you can make a PR now. I think this will make it easier to merge the next steps you mention, and also help @taeoukkim develop the 1D coupling.

**Comment by javijv4** [comment]
@taeoukkim, that's great. Just a heads-up: the last PR (about the Integrator Class) changed a lot of things, which caused some conflicts. I think, before anything else, you should get your branch up to date with the latest version in my branch, which resolves all these conflicts. Happy to chat if any problems arise. 
Then, I think once the PR I'll submit now is merged, you can submit another PR to add the 1D coupling. Does that sound good?

**Comment by taeoukkim** [comment]
@javijv4 Okay, I tryed to rebase my code on your latest commit and yes, it already makes error even when I run simple fluid problem. I will try to resolve codes.

**Comment by javijv4** [comment]
I just submitted the PR. I wanted to add here some more information regarding a specific aspect of the implementation that I haven't discussed previously.

I removed the legacy implementation of the svZeroD coupling, which used the `cplBC` object to store the required information in each face through `cplBC.fa`. So, for example, in `svZeroD_interface.cpp` (previously `svZeroD_subroutines.cpp` there is no mention of `cplBC.fa`. However, I am using `cplBC` to store "global" information about the coupled boundary conditions, for example, the total number of `CoupledBoundaryCondition` objects. 

Note that `cplBC.fa` is still being used for GenBC and RCR BC. So you can still see `cplBC.fa` calls in `set_bc.cpp`, for example. 

Unrelated to the PR, but related to this comment and future implementation:
As discussed before, ideally, the GenBC and RCR BC functionality should be moved to the class. My initial idea (for the future) is to keep using `cplBC` as an object that is aware of all the coupled boundary conditions. For example, it will know how many RCR and svZeroD BCs there are, but it will not store information or be involved in any operation required for each BC; that will happen in the class. We can discuss this in more detail when I actually start working on merging all the coupled BCs, but this is what I was thinking of eventually getting to in the current implementation.

**Comment by taeoukkim** [comment]
Completed the development of the 3D–1D coupling implementation.
Current version of the code can do:
(1) 3D–1D coupling for both Dirichlet and Neumann boundary conditions
(2) RCR boundary condition support with 0D/1D coupling interfaces
(3) Dirichlet-type 3D–0D coupling implemented

Details: https://github.com/SimVascular/svMultiPhysics/issues/512

**Comment by aabrown100-git** [comment]
Great work @taeoukkim! Let's get @javijv4's PR merged (should be done in the next couple days), then we can merge your features. I think it might be better to split your work into 3 separate PR's. If it's not too much work, could you set up your repository for that.
